ledger-service/http-json/src/main/scala/com/digitalasset/http/endpoints/MeteringReportEndpoint.scala

```diff
@@ -0,0 +1,199 @@
+    jsonFormat2(ApplicationMeteringReport.apply)
+
+  implicit val MeteringReportFormat: RootJsonFormat[MeteringReport] =
+    jsonFormat4(MeteringReport.apply)
+
+  private[endpoints] def toTimestamp(pbTimestamp: protobuf.timestamp.Timestamp): Timestamp = {
+    Timestamp.assertFromInstant(
+      Instant.ofEpochSecond(pbTimestamp.seconds, pbTimestamp.nanos.toLong)
+    )
+  }
+
+  private val startOfDay = LocalTime.of(0, 0, 0)
+  private[endpoints] def toTimestamp(ts: LocalDate): Timestamp = {
+    Timestamp.assertFromInstant(Instant.ofEpochSecond(ts.toEpochSecond(startOfDay, ZoneOffset.UTC)))
+  }
+
+  private[endpoints] def toPbTimestamp(ts: Timestamp): protobuf.timestamp.Timestamp = {
+    val instant = ts.toInstant
+    protobuf.timestamp.Timestamp(instant.getEpochSecond, instant.getNano)
+  }
+
+  private[endpoints] def toPbRequest(
+      request: MeteringReportRequest
+  ): metering_report_service.GetMeteringReportRequest = {
+    import request._
+    metering_report_service.GetMeteringReportRequest(
+      from = Some(toPbTimestamp(from)),
+      to = to.map(toPbTimestamp),
+      applicationId = application.fold(
+        metering_report_service.GetMeteringReportRequest.defaultInstance.applicationId
+      )(identity),
+    )
+  }
+
+  // With Protobuf optional string is represented as an empty strings
+  private def pbOption(str: String): Option[String] = Option(str).filter(_.nonEmpty)
+
+  private def mustHave[T](option: Option[T], field: String): Either[String, T] = option match {
+    case Some(t) => Right(t)
+    case None => Left(s"GetMeteringReportResponse missing field, expected $field")
+  }
+
+  type StringOr[T] = Either[String, T]
+
+  private[endpoints] def toMeteringReport(
+      pbResponse: metering_report_service.GetMeteringReportResponse
+  ): Error \/ MeteringReport = {
+
+    (for {
+      pbParticipantReport <- mustHave(pbResponse.participantReport, "participantReport")
+      participantId <- Ref.ParticipantId.fromString(pbParticipantReport.participantId)
+      pbRequest <- mustHave(pbResponse.request, "request")
+      applicationId <- pbOption(pbRequest.applicationId).traverse[StringOr, Ref.ApplicationId](
+        Ref.ApplicationId.fromString
+      )
+      pbFrom <- mustHave(pbRequest.from, "from")
+      request = MeteringReportRequest(
+        from = toTimestamp(pbFrom),
+        to = pbRequest.to.map(toTimestamp),
+        application = applicationId,
+      )
+      applications <- pbParticipantReport.applicationReports
+        .traverse[StringOr, ApplicationMeteringReport] { r =>
+          ApplicationId.fromString(r.applicationId).map { app =>
+            ApplicationMeteringReport(app, r.eventCount)
+          }
+        }
+    } yield MeteringReport(
+      participantId,
+      request,
+      pbParticipantReport.isFinal,
+      applications,
+    )) match {
+      case Right(report) => \/-(report)
+      case Left(message) => -\/(ServerError(message))
+    }
```

Comment:
Still didn't work.  Probably a silly mistake on my side due to not using `disjunction` but it is for that reason I prefer the explicit mapping anyhow.


![Screenshot 2022-03-09 at 11 24 06](https://user-images.githubusercontent.com/56595114/157432430-8ff2eccb-d4cd-4d8a-b48b-f2a57fd96d66.png)

Reply:
Since this PR and an incompatible one got merged concurrently I had to fix this piece of code to make it compile with the incompatible changes to `ServerError`. [I amended the code](https://github.com/digital-asset/daml/pull/13224/files) to match @S11001001's point. I believe your IJ was not type-checking correctly. Unfortunately libraries like `scalaz` may easily confuse IJ.

Reply:
How annoying, thanks for sorting.